libraries/chain/include/eosio/chain/types.hpp

```diff
@@ -410,20 +409,78 @@ namespace eosio::chain {
+   template<typename Stream>
+   inline void pack(Stream& s, const chainbase::shared_string& v) {		
+      FC_ASSERT(v.size() <= MAX_SIZE_OF_BYTE_ARRAYS);		
+      fc::raw::pack(s, fc::unsigned_int((uint32_t)v.size()));		
+      if (v.size())
+         s.write(v.data(), v.size());		
+   }		
+
+   template<typename Stream>
+   inline void unpack(Stream& s, chainbase::shared_string& v) {
+      fc::unsigned_int sz;
+      fc::raw::unpack(s, sz);
+      if (sz) {
+         v.resize_and_fill(sz, [&](char* buf, std::size_t sz){
+            s.read(buf, sz);            
+         });
+      }
+   }
+   
+   // pack/unpack chainbase::shared_cow_vector
+   // ----------------------------------------
+   template<typename Stream, typename T>
+   inline void pack(Stream& s, const chainbase::shared_cow_vector<T>& v)  {
+      FC_ASSERT(v.size() <= MAX_SIZE_OF_BYTE_ARRAYS);
+      fc::raw::pack( s, fc::unsigned_int((uint32_t)v.size()));
+      for (const auto& el : v)
+         fc::raw::pack(s, el);
+   }
+
+   template<typename Stream, typename T>
+   inline void unpack(Stream& s, chainbase::shared_cow_vector<T>& v)  {
+      fc::unsigned_int size;
+      fc::raw::unpack( s, size );
+      FC_ASSERT(size.value <= MAX_SIZE_OF_BYTE_ARRAYS);
+      FC_ASSERT(v.size() == 0);
+      v.clear_and_construct(size.value, 0, [&](auto* dest, std::size_t i) { 
+         new (dest) T(); // unpack expects a constructed variable
+         fc::raw::unpack(s, *dest);
+      });
+   }
+}
+
 namespace chainbase {
    // chainbase::shared_cow_string
-   template<typename DataStream> inline DataStream& operator<<( DataStream& s, const chainbase::shared_cow_string& v )  {
-      FC_ASSERT( v.size() <= MAX_SIZE_OF_BYTE_ARRAYS );
-      fc::raw::pack( s, fc::unsigned_int((uint32_t)v.size()));
-      if( v.size() ) s.write( v.data(), v.size() );
+   // ----------------------------
+   template<typename Stream>
+   inline Stream& operator<<(Stream& s, const chainbase::shared_cow_string& v)  {
+      fc::raw::pack(s, v);
+      return s;
+   }
+
+   template<typename Stream>
+   inline Stream& operator>>(Stream& s, chainbase::shared_cow_string& v)  {
+      fc::raw::unpack(s, v);
+      return s;
+   }
+
+   // chainbase::shared_cow_vector
+   // ----------------------------
+   template<typename Stream, typename T>
+   inline Stream& operator<<(Stream& s, const chainbase::shared_cow_vector<T>& v)  {
+      fc::raw::pack(s, v);
       return s;
    }
 
-   template<typename DataStream> inline DataStream& operator>>( DataStream& s, chainbase::shared_cow_string& v )  {
-      fc::unsigned_int size; fc::raw::unpack( s, size );
-      FC_ASSERT( size.value <= MAX_SIZE_OF_BYTE_ARRAYS );
-      FC_ASSERT( v.size() == 0 );
-      v.resize_and_fill(size.value, [&s](char* buf, std::size_t sz) { s.read(buf, sz); });
+   template<typename Stream, typename T>
+   inline Stream& operator>>(Stream& s, chainbase::shared_cow_vector<T>& v)  {
+      fc::raw::unpack(s, v);
       return s;
    }
 }
```

Comment:
I think all of this should be moved to `database_utils.hpp`

Reply:
done